### PR TITLE
MoE

### PR DESCRIPTION
**Pull request by vchiley**
init experimental moe branch

~~requirements file known issues:~~
```
mosaicml @ git+https://github.com/abhi-mosaic/composer.git@abhi/fsdp_process_group
tutel @ git+https://github.com/vchiley/tutel@main
```

~~ideally `mosaicml @ git+https://github.com/abhi-mosaic/composer.git@abhi/fsdp_process_group` is merged into composer dev and subsequently part of the next release~~
~~Similarly, it'd be nice if `tutel @ git+https://github.com/vchiley/tutel@main` was [merged into tutel main](https://github.com/microsoft/tutel/pull/194) or the [mosaicml org](https://github.com/mosaicml) owned the tutel fork~~

### CODE REVIEW

**Comment by dblalock** [comment]
for context, will this be ready for customers to use by the next release? Asking because we should probably put it in the private fork if not, since the public repo is basically production code.

**Comment by vchiley** [comment]
This isn't being merged to `main`; this is the creation of an experimental moe branch.
This enables "super users" and "research customers" to use moe with the [caveat that this is a experimental branch](https://github.com/mosaicml/examples/pull/180/files#diff-7174cda7646e59e66a2fc56f1f8463af770530542c6987bdae29334f09ecee3aR8)
When we think its "production ready", we'll merge it into main.

**Comment by dblalock** [comment]
So I noticed this grew from 692 lines to 2600 lines; is this just a matter of mosaicml:moe not being updated with what's in main? I started to review it with the added size but it's hard to tell what is and isn't moe-related

**Comment by vchiley** [comment]
merged main into mosaicml:moe

**Review by dblalock — APPROVED** [review]
Just a bunch of minor local stuff.

Also, for reference, I'm trying to review everything in this repo with pedagogy set to maximum in order to contribute to our collective SWE skill--so comments aren't a reflection of thinking you don't know something, just me trying to always err on the side of explaining stuff.

File: examples/llm/tests/test_moe.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+
+def check_tensor(tensor):
+    world_size = dist.get_world_size()
+    tensors = dist.all_gather(tensor)
+    for i in range(world_size):
+        for j in range(i + 1, world_size):
+            assert not tensors[i].equal(tensors[j])
+            if tensors[i].all_close(tensors[j]):
+                warnings.warn(f'')
```
[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/tests/test_moe.py:61]: I need to add an actual warning here.
If parameters are sync'd then `assert not tensors[i].equal(tensors[j])` will catch this as an issue.
If parameters are not sync'd but gradients are then eventually parameters will begin to be the similar over time. Which is why I make it into a warning; in general they're allowed to be close, but it might be indicative of an issue.

[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/tests/test_moe.py:61]: rm

File: examples/llm/src/models/layers/moe/base_moe.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+import torch.nn as nn
+from composer.utils import dist, get_device
+from torch.distributed import new_group
+
+
+class BaseMoE(nn.Module):
```
[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/src/models/layers/moe/base_moe.py:9]: sure

File: examples/llm/src/models/layers/moe/tutel_moe.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                 block_idx: int,
+                 device: Optional[str] = None):
+        super().__init__()
+
+        world_size = dist.get_world_size()
+
+        num_experts = cfg.moe.get('num_experts')
+        if isinstance(num_experts, ListConfig):
+            # enables pyramid moe
+            num_experts = num_experts[block_idx]
```
[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/src/models/layers/moe/tutel_moe.py:37]: PyramidMoE is implemented, and the one test I had runs a PyramidMoE.
I'll add a variant that runs a regular moe.

[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/src/models/layers/moe/tutel_moe.py:37]: Would the right way to test for behavior not impl for PyramidMoEs, be to somehow verify that the MoE lib being used doesn't require that at every depth the user inits the same number of experts?
This sounds heavily based on MoE lib implementation.

[the remark this answers is not in the document]
Comment (dblalock) [reply examples/llm/src/models/layers/moe/tutel_moe.py:37]: I was just thinking it would be nice to just try using num_experts has `__getitem__` rather than checking that it's a particular class because "[depend on abstractions, not concretions](https://softwareengineering.stackexchange.com/questions/401769/depend-on-abstractions-not-on-concretions-what-is-the-exact-meaning-of-this-t)". Something like:
```suggestion
        if isinstance(num_experts, Sequence):
            # enables pyramid moe
            num_experts = num_experts[block_idx]
```
which counts since `isinstance(..., Sequence)` tests for an interface, not a concrete class. Or, using the principle that in Python it's [better to ask for forgiveness than permission](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/), we could:
```python
try:
    num_experts = num_experts[block_idx]
except TypeError:
    pass
```
although personally I think this is clunkier.

File: examples/llm/tests/test_moe.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+from examples.common.builders import (build_algorithm, build_callback,
+                                      build_dataloader, build_logger,
+                                      build_optimizer, build_scheduler)
+from examples.common.config_utils import log_config
+from examples.llm.src.model_registry import COMPOSER_MODEL_REGISTRY
+
+
+def gpt_tiny_moe_cfg(conf_path='yamls/mosaic_gpt/125m_moe.yaml', pyramid=False):
+    """Create gpt tiny moe cfg."""
```
Comment (bcui19) [inline examples/llm/tests/test_moe.py:22]: At some point we might want to consider creating a smaller moe to test, esp if we move to github GPU tests (rather than local?)

File: examples/llm/src/models/layers/gpt_blocks.py [hunk 5]
```diff
@@ -32,14 +35,32 @@ class GPTBlock(nn.Module):
...
             assert cfg.attn_impl == 'triton' or cfg.attn_impl == 'torch', 'Only triton kernel or torch supports alibi'
         self.ln_1 = nn.LayerNorm(cfg.d_model, device=device)
         self.causal_attn = causal_attn_cls(cfg, device)
         self.ln_2 = nn.LayerNorm(cfg.d_model, device=device)
-        self.mlp = GPTMLP(cfg, device=device)
+
+        use_moe = False
+        if cfg.get('moe', None):
+            num_experts = cfg.moe.get('num_experts')
+            if isinstance(num_experts, Sequence):
```
Comment (bcui19) [inline examples/llm/src/models/layers/gpt_blocks.py:50]: Do we want to raise an error here if it is a sequence and `num_experts` doesn't match the `num_blocks`?

**Review by bcui19 — APPROVED** [review]
Very minor comments, but LGTM!!

File: examples/llm/tests/test_moe.py (hunk 4 above)
Comment (vchiley) [reply examples/llm/tests/test_moe.py:24]: yeah rn the test goes through quite a bit and realistically only tests things when on a multi-GPU, multi-node setup.
(I'm not sure GitHub will enable that; maybe it does, just not sure)
